crates/next-core/src/next_server/resolve.rs

```diff
@@ -0,0 +1,98 @@
+use anyhow::Result;
+use turbo_tasks::primitives::StringsVc;
+use turbo_tasks_fs::{glob::GlobVc, FileJsonContent, FileSystemPathVc};
+use turbopack_core::resolve::{
+    find_context_file, package_json,
+    parse::RequestVc,
+    plugin::{ResolvePlugin, ResolvePluginConditionVc, ResolvePluginVc, ResolveResultOptionVc},
+    FindContextFileResult, ResolveResult, SpecialType,
+};
+
+#[turbo_tasks::value]
+pub(crate) struct ExternalCjsModulesResolvePlugin {
+    root: FileSystemPathVc,
+    transpiled_packages: StringsVc,
+}
+
+#[turbo_tasks::value_impl]
+impl ExternalCjsModulesResolvePluginVc {
+    #[turbo_tasks::function]
+    pub fn new(root: FileSystemPathVc, transpiled_packages: StringsVc) -> Self {
+        ExternalCjsModulesResolvePlugin {
+            root,
+            transpiled_packages,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl ResolvePlugin for ExternalCjsModulesResolvePlugin {
+    #[turbo_tasks::function]
+    fn condition(&self) -> ResolvePluginConditionVc {
+        ResolvePluginConditionVc::new(self.root, GlobVc::new("**/node_modules"))
+    }
+
+    #[turbo_tasks::function]
+    async fn after_resolve(
+        &self,
+        fs_path: FileSystemPathVc,
+        _request: RequestVc,
+    ) -> Result<ResolveResultOptionVc> {
+        // always bundle transpiled modules
+        let transpiled_glob = packages_glob(self.transpiled_packages).await?;
+        if transpiled_glob.execute(&fs_path.await?.path) {
+            return Ok(ResolveResultOptionVc::none());
+        }
+
+        let FindContextFileResult::Found(package_json, _) =
+            *find_context_file(fs_path.parent(), package_json()).await?
+        else {
+            // can't find package.json
+            return Ok(ResolveResultOptionVc::none());
+        };
+        let FileJsonContent::Content(package) = &*package_json.read_json().await? else {
+            // can't parse package.json
+            return Ok(ResolveResultOptionVc::none());
+        };
```

Comment:
This could be a single `find_and_read_package_json` method that could be reused in `crates/turbopack-core/src/resolve/mod.rs`.

Reply:
Was looking into it, but because I wanted both path and content it would require a custom return type

matching on the `FileJsonContent::Content` is also only possible here because `package` is a `&serde_json::Value` (can't return that directly)